internal/cmd/generateinterceptor/main.go

```diff
@@ -0,0 +1,412 @@
+		err := s.encode(req)
+		if err != nil {
+			return err
+		}
+
+		err = invoker(ctx, method, req, response, cc, opts...)
+		if err != nil {
+			return err
+		}
+
+		return s.decode(response)
+	}, nil
+}
+`
+
+var codeTemplate = template.Must(template.New("process").Parse(codeTemplateText))
+
+// TypeRecord holds the state for a type referred to by the workflow service
+type TypeRecord struct {
+	Methods []string // List of methods on this type that can eventually lead to Payload(s)
+	Slice   bool     // The API refers to slices of this type
+	Map     bool     // The API refers to maps with this type as the value
+	Matches bool     // We found methods on this type that can eventually lead to Payload(s)
+}
+
+var records = map[string]*TypeRecord{}
+
+// isSlice returns true if a type is slice, false otherwise
+func isSlice(t types.Type) bool {
+	switch t.(type) {
+	case *types.Slice:
+		return true
+	}
+	return false
+}
+
+// isMap returns true if a type is map, false otherwise
+func isMap(t types.Type) bool {
+	switch t.(type) {
+	case *types.Map:
+		return true
+	}
+	return false
+}
+
+// elemType returns the elem (value) type for a slice or map
+func elemType(t types.Type) types.Type {
+	switch typ := t.(type) {
+	case *types.Slice:
+		return typ.Elem()
+	case *types.Map:
+		return typ.Elem()
+	}
+	return t
+}
+
+// typeName returns a normalized path for a type
+func typeName(t types.Type) string {
+	return types.TypeString(elemType(t), func(p *types.Package) string {
+		return p.Name() + "pb"
+	})
+}
+
+// typeMatches returns true if a type:
+// Is equal to or is a pointer to any of the desired types
+// Is a slice or slice of pointers to any of the desired types
+// Is a map where the value is any of the desired types or a pointer to any of the desired types
+func typeMatches(t types.Type, desired ...types.Type) bool {
+	resolved := resolveType(t).String()
+	for _, f := range desired {
+		if resolved == f.String() {
+			return true
+		}
+	}
+
+	return false
+}
+
+// resolveType returns the underlying type for pointers, slices and maps
+func resolveType(t types.Type) types.Type {
+	switch typ := t.(type) {
+	case *types.Pointer:
+		return resolveType(typ.Elem())
+	case *types.Slice:
+		return resolveType(typ.Elem())
+	case *types.Map:
+		return resolveType(typ.Elem())
+	}
+
+	return t
+}
+
+func pruneRecords(input map[string]*TypeRecord) map[string]*TypeRecord {
+	result := map[string]*TypeRecord{}
+
+	for typ, record := range input {
+		if record.Matches {
+			result[typ] = record
+		}
+	}
+
+	return result
+}
+
+// walk iterates the methods on a type and returns whether any of them can eventually lead to Payload(s)
+// The return type for each method on this type is walked recursively to decide which methods can lead to Payload(s)
+func walk(desired []types.Type, skip []types.Type, typ types.Type) bool {
```

Comment:
You'll want to do a few things to make this reusable:

* Probably accept a stateful object as a param that you can build up instead of just bools
* Move this function outside of cmd (but still in a child package of internal) so it can be used by tests
* During walk, collect a set of payload paths that we do want to handle
* Serialize that set of payload paths into a variable in the generated code
* Write a test that does the walk and compares the resulting payload paths to the serialized variable

Also, there are technically other ways you could code-gen based on these paths (e.g. a map of collectors by message type that add payloads to a large set to be encoded/decoded at the end), but what you have is fine too.